Sources/Purchasing/StoreKit2/Observer Mode/StoreKit2ObserverModePurchaseDetector.swift

```diff
@@ -0,0 +1,109 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  StoreKit2ObserverModePurchaseDetector.swift
+//
+//  Created by Will Taylor on 5/1/24.
+
+import Foundation
+import StoreKit
+
+/// Protocol describing an actor capable of detecting purchases from StoreKit 2.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+protocol SK2ObserverModePurchaseDetectorType {
+    func detectUnobservedTransactions(
+        delegate: StoreKit2ObserverModeManagerDelegate?
+    ) async
+}
+
+/// Actor responsibile for detecting purchases from StoreKit2 that should be processed by observer mode.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+actor StoreKit2ObserverModePurchaseDetector: SK2ObserverModePurchaseDetectorType {
+
+    private let deviceCache: DeviceCache
+    private let currentUserProvider: CurrentUserProvider
+    private let allTransactionsProvider: AllTransactionsProviderType
+
+    init(
+        deviceCache: DeviceCache,
+        currentUserProvider: CurrentUserProvider,
+        allTransactionsProvider: AllTransactionsProviderType
+    ) {
+        self.deviceCache = deviceCache
+        self.currentUserProvider = currentUserProvider
+        self.allTransactionsProvider = allTransactionsProvider
+    }
+
+    /// Detects unobserved transactions and forwards them to the StoreKit2ObserverModeManagerDelegate
+    /// for processing.
+    func detectUnobservedTransactions(
+        delegate: StoreKit2ObserverModeManagerDelegate?
+    ) async {
+        guard let mostRecentVerifiedTransaction: (
+            verifiedTransaction: StoreKit.Transaction,
+            jwsRepresentation: String
+        ) = (await allTransactionsProvider.getAllTransactions()
+            .compactMap { transaction in
+                guard let verifiedTransaction = transaction.verifiedTransaction else {
+                    return nil
+                }
+                return (
+                    verifiedTransaction: verifiedTransaction,
+                    jwsRepresentation: transaction.jwsRepresentation
+                )
+            }
+            .sorted(by: { $0.verifiedTransaction.purchaseDate > $1.verifiedTransaction.purchaseDate })
+            .first
+        ) else {
+            return
+        }
```

Comment:
I'll be the first to admit that I had to read this a couple of times because I'm not used to the mix of very idiomatic swift features, from the tuple to the chain of functional calls afterwards

Reply:
maybe we can dumb it down or move the tuple to a named struct, or just do one thing at a time?

Reply:
Not a problem! I've changed it so it all happens in discrete steps (and got rid of the tuple!) 😄